### ISSUE TITLE
Allow page template outside of /site/templates (inside of module path, module generated page) or use process modules in frontend (parent path "/" instead of admin)

### ISSUE THREAD

**Issue by pwFoo** [processwire/processwire-requests#204]
Hi @ryancramerdesign 
At the moment I try to get a page generated by a module working without template in `site/templates`. The template shouldn't be modified and is located in the `site/modules/<MODULE>/templates/myTemplate.php`.

Because it isn't possible I tried to change the module to a process module, because the page path have to be "/" as parent which seems not to work (404 error because template not found) with process modules.

A workaround could be to copy the module template to `site/templates` during module install / upgrade and delete the template if the module is uninstalled.
But I would prefer a better solution (frontend process module or template inside of module path).

**Comment by Toutouwai** [comment]
It is possible to use a template file outside of /site/templates/. Examples of how you can set the template filename...

In /site/init.php

```php
$t = $templates->get('my_template');
$t->filename = $config->paths->MyModule . 'my_template.php';
```

That would set the filename for the template every time ProcessWire initialises. If you wanted to be more selective you could use a hook in /site/init.php:

```php
$wire->addHookBefore('Page::viewable', function(HookEvent $event) {
    $page = $event->object;
    if($page->template != 'my_template') return;
    $page->template->filename = $this->config->paths->MyModule . 'my_template.php';
});
```

But for either of these to work you must disable FileCompiler for the template on the Files tab of Edit Template. That's because FileCompiler forces the $config->paths->templates path at the start of the filename you set for the template, resulting in an invalid path if you used a full path outside of the templates folder for the filename. Related code:

https://github.com/processwire/processwire/blob/341342dc5b1c58012ae7cb26cffe2c57cd915552/wire/modules/PageRender.module#L402
https://github.com/processwire/processwire/blob/341342dc5b1c58012ae7cb26cffe2c57cd915552/wire/modules/PageRender.module#L499-L503
https://github.com/processwire/processwire/blob/341342dc5b1c58012ae7cb26cffe2c57cd915552/wire/core/FileCompiler.php#L339-L344

@ryancramerdesign, could FileCompiler be changed to allow template files outside of /site/templates?

**Comment by pwFoo** [comment]
Thanks @Toutouwai ,
I know the workarounds with additional code in init.php or a hook. But I try to save and use the template inside of the modules directory to avoid workarounds outside of the module directory (autoload module with hook, additional code in init.php, copy modules template to site/templates).

A nice solution would be to set the needed template file in modules install() method and save the custom template path to the page / template.

**Comment by jlahijani** [comment]
> ```
> $t = $templates->get('my_template');
> $t->filename = $config->paths->MyModule . 'my_template.php';
> ```

Did this technique break with PW 3.0.141?  I'm getting a white screen of death. @Toutouwai

**Comment by Toutouwai** [comment]
@jlahijani, yes, it seems this got broken by the change below, although no white screen for me - rather setting the custom filename just doesn't work. https://github.com/processwire/processwire/commit/710c222b5a964dcc804542c33a7d325e53f50c55#diff-bf77104b4984638acb422b9e9b6739aeR899-R903

This change prevents setting a custom filename at runtime. Not sure why this has been done, and it seems to contradict the [phpDoc comment](https://github.com/processwire/processwire/blob/3c11bbf1d87ba6ccfe0a44ad5c2ebecaf865cc5f/wire/core/Template.php#L77) for the filename property:

> @property string $filename Template filename, including path (this is auto-generated from the name, though you may modify it at runtime if it suits your need).
